plugins/dell-k2/fu-dell-k2-ec.c

```diff
@@ -672,75 +663,160 @@ fu_dell_k2_ec_commit_package(FuDevice *device, GBytes *blob_fw, GError **error)
+				    GError **error)
+{
+	gsize fw_sz = 0;
+	gsize chunk_sz = 0;
+	gint first_packet_delay_time = 0;
 	g_autoptr(GBytes) fw = NULL;
-	g_autoptr(GBytes) fw_whdr = NULL;
 	g_autoptr(FuChunkArray) chunks = NULL;
 
+	/* basic test */
 	g_return_val_if_fail(device != NULL, FALSE);
 	g_return_val_if_fail(FU_IS_FIRMWARE(firmware), FALSE);
 
-	/* progress */
-	fu_progress_set_id(progress, G_STRLOC);
-	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_READ, 1, NULL);
-	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_ERASE, 12, NULL);
-	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 85, NULL);
-	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_VERIFY, 1, NULL);
-
-	if (!fu_dell_k2_ec_hid_raise_mcu_clock(device, TRUE, error))
-		return FALSE;
-	fu_progress_step_done(progress);
-
 	/* get default image */
 	fw = fu_firmware_get_bytes(firmware, error);
 	if (fw == NULL)
 		return FALSE;
 
-	/* construct writing buffer */
-	fw_whdr = fu_dell_k2_ec_hid_fwup_pkg_new(fw, DELL_K2_EC_DEV_TYPE_MAIN_EC, 0);
+	/* payload size */
+	fw_sz = g_bytes_get_size(fw);
+
+	/* adjust the chunk size by device type */
+	switch (dev_type) {
+	case DELL_K2_EC_DEV_TYPE_MAIN_EC:
+		chunk_sz = DELL_K2_EC_DEV_EC_CHUNK_SZ;
+		break;
+	case DELL_K2_EC_DEV_TYPE_RMM:
+		chunk_sz = DELL_K2_EC_DEV_NO_CHUNK_SZ;
+		first_packet_delay_time = 75 * 1000;
+		break;
+	default:
+		chunk_sz = DELL_K2_EC_DEV_ANY_CHUNK_SZ;
+	}
 
-	/* prepare the chunks */
-	chunks = fu_chunk_array_new_from_bytes(fw_whdr,
+	/* maximum buffer size */
+	chunks = fu_chunk_array_new_from_bytes(fw,
 					       FU_CHUNK_ADDR_OFFSET_NONE,
 					       FU_CHUNK_PAGESZ_NONE,
-					       DELL_K2_EC_HID_DATA_PAGE_SZ);
+					       chunk_sz);
 
-	/* erase */
-	if (!fu_dell_k2_ec_hid_erase_bank(device, 0xff, error))
-		return FALSE;
-	fu_progress_step_done(progress);
-
-	/* write to device */
+	/* iterate the chunks */
 	for (guint i = 0; i < fu_chunk_array_length(chunks); i++) {
 		g_autoptr(FuChunk) chk = NULL;
+		g_autoptr(FuChunkArray) pages = NULL;
+		g_autoptr(GBytes) buf = NULL;
+		guint8 res[3] = {0x00};
 
 		chk = fu_chunk_array_index(chunks, i, error);
 		if (chk == NULL)
 			return FALSE;
 
-		if (!fu_dell_k2_ec_hid_write(device, fu_chunk_get_bytes(chk), error))
-			return FALSE;
-	}
-	fu_progress_step_done(progress);
+		/* prepend header and command to the chunk data */
+		buf = fu_dell_k2_ec_hid_fwup_pkg_new(fu_chunk_get_bytes(chk),
+						     fw_sz,
+						     dev_type,
+						     dev_identifier);
+
+		/* slice the chunk into pages */
+		pages = fu_chunk_array_new_from_bytes(buf,
+						      FU_CHUNK_ADDR_OFFSET_NONE,
+						      FU_CHUNK_PAGESZ_NONE,
+						      DELL_K2_EC_HID_DATA_PAGE_SZ);
+
+		/* iterate the pages */
+		for (guint j = 0; j < fu_chunk_array_length(pages); j++) {
+			guint8 packet[DELL_K2_EC_HID_DATA_PAGE_SZ] = {0x00};
+			g_autoptr(FuChunk) page = NULL;
+
+			page = fu_chunk_array_index(pages, j, error);
+			if (page == NULL)
+				return FALSE;
```

Comment:
if you do `fu_chunk_get_bytes()` you could use `fu_bytes_pad()` here I think.

Reply:
using `fu_bytes_pad()` with size `DELL_K2_EC_HID_DATA_PAGE_SZ` it is complaing a stack overflow issue.

```
*** stack smashing detected ***: terminated
```